internal/installercache/installercache.go

```diff
@@ -1,61 +1,187 @@
 	var err error
-	//cache miss
-	if r.path == "" {
-		mirrorRegistriesBuilder := mirrorregistries.New()
-		path, err = oc.NewRelease(&executer.CommonExecuter{}, oc.Config{
-			MaxTries: oc.DefaultTries, RetryDelay: oc.DefaltRetryDelay}, mirrorRegistriesBuilder).Extract(log, releaseID, releaseIDMirror, cacheDir, pullSecret, platformType, icspFile)
+
+	workdir, binary, path = ocRelease.GetReleaseBinaryPath(releaseID, i.cacheDir, platformType)
+	if _, err = os.Stat(path); os.IsNotExist(err) {
+		//evict older files if necessary
+		i.evict()
+
+		//extract the binary
+		_, err = ocRelease.Extract(i.log, releaseID, releaseIDMirror, i.cacheDir, pullSecret, platformType, icspFile)
+		if err != nil {
+			return &Release{}, err
+		}
+	} else {
+		//update the file mtime to signal it was recently used
+		err = os.Chtimes(path, time.Now(), time.Now())
+		if err != nil {
+			return &Release{}, errors.Wrap(err, fmt.Sprintf("Failed to update release binary %s", path))
+		}
+	}
+	// return a new hard link to the binary file
+	// the caller should delete the hard link when
+	// it finishes working with the file
+	link := filepath.Join(workdir, "ln_"+fmt.Sprint(time.Now().Unix())+
+		"_"+binary)
+	err = os.Link(path, link)
+	if err != nil {
+		return &Release{}, errors.Wrap(err, fmt.Sprintf("Failed to create hard link to binary %s", path))
+	}
+	return &Release{link}, nil
+}
+
+//	Walk through the cacheDir and list the files recursively.
+// If the total volume of the files reaches the capacity, delete
+// the oldest ones.
+//
+// Locking must be done outside evict() to avoid contentions.
+func (i *Installers) evict() {
+	//if cache limit is undefined skip eviction
+	if i.storageCapacity == 0 {
+		return
+	}
+
+	// store the file paths
+	files := NewPriorityQueue(&fileInfo{})
+	links := make([]*fileInfo, 0)
+	var totalSize int64
+
+	// visit process the file/dir pointed by path and store relevant
+	// paths in a priority queue
+	visit := func(path string, info os.FileInfo, err error) error {
 		if err != nil {
-			return "", err
+			return err
+		}
+
+		if !info.Mode().IsRegular() {
+			return nil
+		}
+		//find hard links
+		if strings.HasPrefix(info.Name(), "ln_") {
+			links = append(links, &fileInfo{path, info})
+			return nil
 		}
-		r.path = path
+
+		//save the other files based on their mod time
+		files.Add(&fileInfo{path, info})
+		totalSize += info.Size()
+		return nil
+	}
+
+	err := filepath.Walk(i.cacheDir, visit)
+	if err != nil {
+		if !os.IsNotExist(err) { //ignore first invocation where the cacheDir does not exist
+			i.log.WithError(err).Errorf("release binary eviction failed to inspect directory %s", i.cacheDir)
+		}
+		return
+	}
+
+	//prune the hard links just in case the deletion in ignition.go
+	//did not succeeded as expected
+	for idx := 0; idx < len(links); idx++ {
+		finfo := links[idx]
+		grace := time.Now().Add(DeleteGracePeriod).Unix()
+		if finfo.info.ModTime().Unix() < grace {
+			os.Remove(finfo.path)
```

Comment:
Not sure I understand this.

Couldn't this delete something that is currently being used?
Or is that what the grace period is for? We're assuming that it won't be in use for more than 5 minutes?
If that's the case it's probably worth a comment next to the grace period definition.

Reply:
Exactly. we assume that it will not be used more than 5 minutes. I will add a comment